core/src/main/scala/chisel3/Data.scala

```diff
@@ -893,6 +893,50 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
   def toPrintable: Printable
 }
 
+object Data {
+  implicit class DataEquality[T <: Data](lhs: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions) {
+    def ===(rhs: T): Bool = {
+      (lhs, rhs) match {
+        case (thiz: UInt, that: UInt) => thiz === that
+        case (thiz: SInt, that: SInt) => thiz === that
+        case (thiz: AsyncReset, that: AsyncReset) => thiz.asBool === that.asBool
+        case (thiz: Reset, that: Reset) => thiz === that
+        case (thiz: Interval, that: Interval) => thiz === that
+        case (thiz: FixedPoint, that: FixedPoint) => thiz === that
+        case (thiz: EnumType, that: EnumType) => thiz === that
+        case (thiz: Clock, that: Clock) => thiz.asUInt === that.asUInt
+        case (thiz: Vec[_], that: Vec[_]) =>
+          if (thiz.length != that.length) {
+            false.B
+          } else {
+            thiz.getElements
+              .zip(that.getElements)
+              .map { case (thisData, thatData) => thisData === thatData }
+              .reduce(_ && _)
+          }
+        case (thiz: Record, that: Record) =>
+          if (thiz.elements.size != that.elements.size) {
+            false.B
+          } else {
+            thiz.elements
+              .zip(that.elements)
```

Comment:
what is the rule on these, do elements have to go in some defined order? What if I have something with `a` and `b` and something with `b` and `a`, shouldn't this work looking stuff up in maps or something vs assumign they are in the same order?

Reply:
In theory by Chisel semantics, we should match the fields by name rather than order (eg. connections work https://scastie.scala-lang.org/HvllDnMUQhS62O77w3vwtA)

So yeah I think this should do something similar to the logic in `typeEquivalent`: https://github.com/chipsalliance/chisel3/blob/1aea4ef96466cbe08150d20c85c88b81e4e4f80f/core/src/main/scala/chisel3/Aggregate.scala#L1094

Unfortunately, this cannot _use_ `typeEquivalent` because it's definition is too strict--it requires that the `Records` have the same runtime class, and it also requires the widths of `UInts` to be the same which is definitely not the case in this function.